File: backend/app/schemas/spotify.py
"""Spotify API response schemas."""

from pydantic import BaseModel
from typing import List, Optional


class SpotifyImage(BaseModel):
    """Spotify image object."""
    url: str
    height: Optional[int] = None
    width: Optional[int] = None


class SpotifyArtistSimple(BaseModel):
    """Simplified artist object."""
    id: str
    name: str
    external_urls: Optional[dict] = None


class SpotifyAlbumSimple(BaseModel):
    """Simplified album object."""
    id: str
    name: str
    images: List[SpotifyImage] = []
    release_date: Optional[str] = None
    external_urls: Optional[dict] = None


class SpotifyUser(BaseModel):
    """Spotify user profile."""
    id: str
    display_name: Optional[str] = None
    email: Optional[str] = None
    images: List[SpotifyImage] = []
    product: Optional[str] = None
    followers: Optional[dict] = None
    external_urls: Optional[dict] = None


class SpotifyArtist(BaseModel):
    """Full Spotify artist object."""
    id: str
    name: str
    genres: List[str] = []
    popularity: Optional[int] = None
    images: List[SpotifyImage] = []
    external_urls: Optional[dict] = None
    followers: Optional[dict] = None


class SpotifyTrack(BaseModel):
    """Spotify track object."""
    id: str
    name: str
    duration_ms: int
    popularity: Optional[int] = None
    album: SpotifyAlbumSimple
    artists: List[SpotifyArtistSimple]
    external_urls: Optional[dict] = None
    preview_url: Optional[str] = None


class SpotifyAlbum(BaseModel):
    """Full Spotify album object with track count."""
    id: str
    name: str
    artists: List[SpotifyArtistSimple]
    images: List[SpotifyImage] = []
    release_date: Optional[str] = None
    total_tracks: Optional[int] = None
    track_count_in_top: int = 0  # Custom field: how many tracks from this album are in user's top
    external_urls: Optional[dict] = None


class TopArtistsResponse(BaseModel):
    """Response for top artists endpoint."""
    items: List[SpotifyArtist]
    total: int
    limit: int
    offset: int
    time_range: str


class TopTracksResponse(BaseModel):
    """Response for top tracks endpoint."""
    items: List[SpotifyTrack]
    total: int
    limit: int
    offset: int
    time_range: str


class TopAlbumsResponse(BaseModel):
    """Response for top albums endpoint (calculated)."""
    items: List[SpotifyAlbum]
    total: int
    limit: int
    time_range: str


class PlayHistoryItem(BaseModel):
    """Item in play history."""
    track: SpotifyTrack
    played_at: str


class RecentlyPlayedResponse(BaseModel):
    """Response for recently played endpoint."""
    items: List[PlayHistoryItem]
    total: int

File: frontend/src/types/spotify.ts
/**
 * Spotify API types
 */

export interface ISpotifyImage {
  url: string
  height: number | null
  width: number | null
}

export interface ISpotifyUser {
  id: string
  display_name: string | null
  email: string | null
  images: ISpotifyImage[]
  product: string | null
  followers: {
    total: number
  } | null
  external_urls: {
    spotify: string
  } | null
}

export interface ISpotifyArtist {
  id: string
  name: string
  genres: string[]
  popularity: number | null
  images: ISpotifyImage[]
  external_urls: {
    spotify: string
  } | null
  followers?: {
    total: number
  } | null
}

export interface ISpotifyArtistSimple {
  id: string
  name: string
  external_urls?: {
    spotify: string
  } | null
}

export interface ISpotifyAlbumSimple {
  id: string
  name: string
  images: ISpotifyImage[]
  release_date: string | null
  external_urls?: {
    spotify: string
  } | null
}

export interface ISpotifyTrack {
  id: string
  name: string
  duration_ms: number
  popularity: number | null
  album: ISpotifyAlbumSimple
  artists: ISpotifyArtistSimple[]
  external_urls: {
    spotify: string
  } | null
  preview_url: string | null
}

export interface ISpotifyAlbum {
  id: string
  name: string
  artists: ISpotifyArtistSimple[]
  images: ISpotifyImage[]
  release_date: string | null
  total_tracks: number | null
  track_count_in_top: number
  external_urls: {
    spotify: string
  } | null
}

export interface ITopArtistsResponse {
  items: ISpotifyArtist[]
  total: number
  limit: number
  offset: number
  time_range: string
}

export interface ITopTracksResponse {
  items: ISpotifyTrack[]
  total: number
  limit: number
  offset: number
  time_range: string
}

export interface ITopAlbumsResponse {
  items: ISpotifyAlbum[]
  total: number
  limit: number
  time_range: string
}

export interface IPlayHistoryItem {
  track: ISpotifyTrack
  played_at: string
}

export interface IRecentlyPlayedResponse {
  items: IPlayHistoryItem[]
  total: number
}

export type TimeRange = 'short_term' | 'medium_term' | 'long_term'

export const TIME_RANGE_LABELS: Record<TimeRange, string> = {
  short_term: 'Ostatni miesiąc',
  medium_term: 'Ostatnie 6 miesięcy',
  long_term: 'Wszystkie czasy',
}
